### ISSUE TITLE
Friends & Family (Zoran)

### ISSUE THREAD

**Issue by kieranbenton** [kaizenticketing/interview-takeawaytask#1]
# Brief
In plain business terms...

* Add a new section to customer's my account to view, add, remove and request permission to 'manage tickets'. Anyone who has allowed us as a purchasing customer to 'manage' their tickets on their behalf should be known as a 'managed' customer.
    * This *also* needs to show a separate list of customers who are 'managing' our account - with the ability for the customer to stop them from acting on their behalf. To 'break' the link.
* Allow prioritisation of these outbound 'links'? TBD.
* Maximum limit of links - 15? Even smaller number of active 'manage tickets' permission.
* Requesting permission to 'manage tickets' should send an email (with digest etc - similar to activate email) of a new notification type to ask the target customer to confirm they want us to manage their tickets
    * Copy the activation email and then iterate with rest of the team on its content
    * There will need to be a landing route/controller to confirm the digest
    * Make sure they are told that they can opt out once this is done through their F&F section
* What if they don't have an email address? On user based channels this should still be allowed without verification? DISCUSS.
* Assign UI should be changed to show a list of linked customers (left) and the existing add / lookup UI on the right
* Change the right to lookup/register then add to the purchasing customer's F&F AND then do the assignment
* Left lets you just click quickly to choose people
* Migrate existing data into this. Any account that was 'created by' another - should be added WITH management permissions?? DISCUSS. Anyone who has otherwise been an attendee where the purchasing customer purchased should be adding as a F&F without 
* Expand getting reservations (wherever that is exposed in the UI) to fetch from the purchasing customer AND all of its 'managed' customers. Need to order, group sensibly (iterate) and allow confirmation of reservations all the way through the flow
* DISCUSS. Should we allow a purchasing customer to view and perform actions on orders (intermingled in their list in a different colour?) of 'managed' customers?
	

# TODO
* ... break down the brief into engineering implementation smaller steps for review ... 

# Database Changes
* ...

    ## Manual Data Updates Required
    * ... in any environment(s) ...

# Setup Notes
* ...

# Test Notes
* ... what areas should QA look at that you might not expect ...

# Out of scope
* ...

# Bugs

## Outstanding

- [x] This is a bug summary
  *This is a repro step 1
  *Step 2
  * :screwdriver: I think this is fixed
  * :raising_hand_man: Need more information
  * :skull_and_crossbones: Not going to fix
  * :art: By Design
  * :boom: Still broken

**Comment by jezdimirovic** [comment]
Hi @kieranbenton 

This is the review I did but now I see it is not complete ofc:

- this would be many to many relation as one customer may have many other custemers who are managing our account/tickets
- about outbound links prio, that should be some kind of (display) order?
- no need for fixed number of links, we sould use table, pagination and filtering
- sending email if we have one, then we try sending SMS (short URL as I had issues with long one)
- not sure about this assign UI, I get the list on left, but what's on the right, details? But what is the change request, do not get the flow with assignment?
- migrating data means we are creating table with F&F where CreatedBy is our current Id?
- if someone added permission to us I assume we could perform all actions

Tasks:
- table with CRUD functionality for F&F
- cancel managing account 
- add/update notification system with SMS
- assignment screen (left/right section)
- migrate task but this need some more details
- reservations

**Comment by kieranbenton** [comment]
Hi @jezdimirovic,
Answers inline.

> * this would be many to many relation as one customer may have many other custemers who are managing our account/tickets

Agreed.

> * about outbound links prio, that should be some kind of (display) order?

Prio = priority? Yes display order, but also potentially to drive logic elsewhere in the purchase flow for example:
- I add 3 tickets to my cart that have a rule set that means each one has to be assigned to a different individual
- Right now - the first ticket is assigned to me, and two are required to be manually assigned
- By adding F&F we could (in theory) auto assign the two tickets to the 'highest' priority connections
- I don't expect us to do this in this phase, but it leaves it open as a possibility for the future

> * no need for fixed number of links, we sould use table, pagination and filtering

Table yes, for me the pagination and filtering are optional (at least initially). We should be encouraging a relatively small number of connections and actually put some kind of limit in place. We should discuss what is reasonable. :-)

> * sending email if we have one, then we try sending SMS (short URL as I had issues with long one)

Definitely email, and (you cant see this yet obviously) we have a templating system and email generation pipeline that can be readily used. SMS we don't have (yet) and is not required initially.

> * not sure about this assign UI, I get the list on left, but what's on the right, details? But what is the change request, do not get the flow with assignment?

Left = ok, yes makes sense. Right should adapt the current implementation - which looks like this:
![image](https://user-images.githubusercontent.com/39090/206689531-fc06394d-59e9-4f06-8131-fa66420f12d6.png)

So the 'right hand side' should still allow someone to assign a ticket to a person (existing or new) without needing to add them to F&F first - we should make it optional (checkbox?) to let someone to their connections at the same time.

> * migrating data means we are creating table with F&F where CreatedBy is our current Id?

So here I mean once that all of the new tables are created and the feature implemented, we should deploy a SQL script as part of our migrations to copy existing customers who were created by others (they were created by using the assign UI in the screenshot above) into new 'connections' with full permissions. Since they in the existing flow are implicitly 'owned' by the creating customer.

> * if someone added permission to us I assume we could perform all actions

Not sure what you mean here - can you clarify?

**Comment by jezdimirovic** [comment]
Hi @kieranbenton 

- Oh, ok, so I get it right about this order, so something like PriorityOrder would be ok for the name of the column
- Still not ok with that limit of links but let's with 10 as in the future, it will be first page of my table :)
- About email, that's mandatory, ofc, but I understand we are discussing if someone is missing email?
- Ok, get the idea now with that image, it's some kind of list with add/edit button. Now I see response for my question above, we have account number OR email address. Well, same here, just add another OR phone nubmer.
- Last point is about if we should allow a purchasing customer to view and perform actions on orders... Not sure how do we handle permissions, if we can create another permission of maybe some claim so we have those micro configurations/permissions.

About #2, well that could be extended in many ways, as you said, I can go wild and do whatever, it is really endless. 🏙️ If you have any particular request, just name it and it will be done.

Thanks

**Comment by kieranbenton** [comment]
> Hi @kieranbenton
> 
> * Oh, ok, so I get it right about this order, so something like PriorityOrder would be ok for the name of the column

Sounds about right. We have a standard name of 'rank' with an associate interface which is likely suitable for this naming wise.

> * Still not ok with that limit of links but let's with 10 as in the future, it will be first page of my table :)

;-) fair enough!

> * About email, that's mandatory, ofc, but I understand we are discussing if someone is missing email?

That's true, but the same could well be true of someone who has no mobile number as well. The common use case for this is something like...
- 'Lead' customer has had an account created in their name by the box office at some point instead of as you might expect directly online. This is common when we take on a new club as we will have imported a lot of customers some of whom may never have bought online before. They have an email address against this account but have never transacted online.
- They finally go online and 'activate' their account (process already exists to do this)
- They realise that they need to be able to link their 2 x children to their account so they can manage their season tickets. Those children accounts exist in the system but never had email addresses against them.
- How can the lead client achieve that? Is it possible at all without manual intervention?

> * Ok, get the idea now with that image, it's some kind of list with add/edit button. Now I see response for my question above, we have account number OR email address. Well, same here, just add another OR phone nubmer.

Makes sense.

> * Last point is about if we should allow a purchasing customer to view and perform actions on orders... Not sure how do we handle permissions, if we can create another permission of maybe some claim so we have those micro configurations/permissions.

I think that only two levels of 'permission' are required
- Standard - can assign new tickets to this person
- Managed - as standard, but can also manage that customer's existing tickets eg. can confirm reservations, reprint/download, transfer and put up for resale their tickets
